File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int diameter = 0;

    public int hieght(TreeNode root) {
        if (root == null) return 0;
        int left = hieght(root.left);
        int right = hieght(root.right);
        int dia = left + right + 1;
        diameter = Math.max(dia, diameter);
        return Math.max(left, right) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        hieght(root);
        System.gc();
        return diameter - 1;
    }
}

File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var result: Int = kotlin.Int.Companion.MIN_VALUE

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        depth(root)
        return result
    }

    fun depth(node: TreeNode?): Int {
        if (node == null) return 0
        val leftDepth = depth(node.left)
        val rightDepth = depth(node.right)

        val currrentD = leftDepth + rightDepth
        result = max(result, currrentD)

        return max(leftDepth, rightDepth) + 1
    }
}